From c6dcc0b136993951dd31502074c3c59c8962acb5 Mon Sep 17 00:00:00 2001
From: Ralph Boehme <slow@samba.org>
Date: Fri, 13 May 2016 12:09:57 +0200
Subject: [PATCH 2/2] afpd: reading from file may fail

When we get a request to list xattrs of a file that:

o is an open fork file handle in use by the client

o has no associated metadata

we call ad_close() on the adouble handle which does a refcount decrement
which closes the file handle.

The fix is remembering whether we the ad_open() actually succeeded and
only call ad_close() in that case.

Bug: https://sourceforge.net/p/netatalk/bugs/619/

Signed-off-by: Ralph Boehme <slow@samba.org>
---
 etc/afpd/extattrs.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/etc/afpd/extattrs.c b/etc/afpd/extattrs.c
index 18f5c78..aee202d 100644
--- a/etc/afpd/extattrs.c
+++ b/etc/afpd/extattrs.c
@@ -87,6 +87,7 @@ int afp_listextattr(AFPObj *obj _U_, char *ibuf, size_t ibuflen _U_, char *rbuf,
 
     static int          buf_valid = 0;
     static size_t       attrbuflen = 0;
+    bool                close_ad = false;
 
     *rbuflen = 0;
     ibuf += 2;
@@ -176,6 +177,7 @@ int afp_listextattr(AFPObj *obj _U_, char *ibuf, size_t ibuflen _U_, char *rbuf,
                 return AFPERR_MISC;
             }
         } else {
+            close_ad = true;
             FinderInfo = ad_entry(adp, ADEID_FINDERI);
             /* Check if FinderInfo equals default and empty FinderInfo*/
             if (memcmp(FinderInfo, emptyFinderInfo, 32) != 0) {
@@ -238,7 +240,7 @@ exit:
     if (ret != AFP_OK)
         buf_valid = 0;
 
-    if (adp)
+    if (close_ad)
         ad_close(adp, ADFLAGS_HF);
 
     return ret;
-- 
2.8.3

